### ISSUE TITLE
Add a new message-action to expand/collapse quoted messages

### ISSUE THREAD

**Issue by Palid** [vector-im/element-web#18884]
We need ability to expand quoted messages like we did previously, so we don't have to jump around with quoted messages to see the entire message's content.

We have settled on the "Tile action bar" design for this one.
Design with different explorations: https://www.figma.com/file/TF6whjf6fHv6mxGjAP7B96/Timeline-Replies?node-id=0%3A1

Follow up from https://github.com/vector-im/element-web/issues/18200

**Comment by Palid** [comment]
Currently the simple solution has a few problems, see https://github.com/matrix-org/matrix-react-sdk/pull/6701

**Comment by janogarcia** [comment]
@Palid Thank you for tacking this one!

I've been thinking more on this issue and ideating how to overcome (or accept) the limitations introduced by this approach that we've been discussing internally.

Some stuff that could be **improved** in the related PR https://github.com/matrix-org/matrix-react-sdk/pull/6701

- Let the replies to expand in full. Do not limit them to an arbitrary height box with scrollbars.
- Change the copy for the tooltips to `Expand replies | ⇧+click` and `Collapse replies | ⇧+click` respectively. Please refer to the [related Figma file](https://www.figma.com/file/TF6whjf6fHv6mxGjAP7B96/Timeline-Replies?node-id=0%3A1) for the design spec. 
- Implement the suggested hotkey functionality if it's not in conflict with any existing feature.

Implementation tradeoffs that should be **acceptable**:

- It's totally fine if we don't track the expanded/collapsed state when switching between rooms for this iteration. 👍 There's limited resources to devote to this, so we need to optimize for the most common use case, which should be expanding and collapsing the replies while staying in the same room. We can improve it later on, in a future iteration, and track the state if really necessary to better serve the less common use case.
- Ideally, the user scrolling position should be preserved when toggling the expanded replies (as it currently happens when part of the contents are beyond the viewport https://cloudup.com/iWHgdoffq10 ), so that the user can expand/collapse repeatedly without seeing the content jump in the timeline and losing focus of the related button.

**Comment by HarHarLinks** [comment]
> Let the replies to expand in full. Do not limit them to an arbitrary height box with scrollbars.

I agree: if you expand a message, then it is more important to be viewed than even the complete timeline.

---

I would like a global setting to determine if default is collapsed or expanded.

**Comment by janogarcia** [comment]
@Palid Updated the ideas above with some additional notes (unsure if you're being notified about those edits by GitHub) on animation, text selection and such.

@HarHarLinks I'll discuss it with the team but, ideally we shouldn't need to introduce a new global setting, and settle on a model that hopefully serves best for the vast majority of users, as those kind of settings can easily add up by introducing more complexity for the users and higher maintenance effort.

**Comment by janogarcia** [comment]
@palid There seems to be a bug with how you determine when a reply is expandable. If the first few lines in a message have only few characters then it won't get the expand/collapse button when it's quoted.

![missing-expand-action](https://user-images.githubusercontent.com/60790/132009636-f6a99fae-df77-47c2-9b4b-89ddc7bb0cc7.png)

**Comment by Palid** [comment]
> @Palid There seems to be a bug with how you determine when a reply is expandable. If the first few lines in a message have only few characters then it won't get the expand/collapse button when it's quoted.
> 
> ![missing-expand-action](https://user-images.githubusercontent.com/60790/132009636-f6a99fae-df77-47c2-9b4b-89ddc7bb0cc7.png)

Thanks for noticing, I'll try to figure out a smarter way to catch it, though it's not really that easy unfortunately.

**Comment by ShadowJonathan** [comment]
@Palid the figma file is inaccessible for the public, btw

**Comment by Palid** [comment]
@janogarcia could you fix the link please? 😅

**Comment by ara4n** [comment]
Some of the figma artboards are kept private to mitigate bikeshedding; if the design team want to keep it that way then that's their prerogative.

**Comment by Palid** [comment]
> > @Palid There seems to be a bug with how you determine when a reply is expandable. If the first few lines in a message have only few characters then it won't get the expand/collapse button when it's quoted.
> > ![missing-expand-action](https://user-images.githubusercontent.com/60790/132009636-f6a99fae-df77-47c2-9b4b-89ddc7bb0cc7.png)
> 
> Thanks for noticing, I'll try to figure out a smarter way to catch it, though it's not really that easy unfortunately.

@janogarcia the thing you reported here is actually a mesh of two different things, one of which is a bug.

1. There's the `expand code block` button, but it doesn't work anymore due to jump-to-quote behavior. I think we should replace the expanding code block in quotes with the expand button for quotes by default, what do you think?
2. Layout is broken for quotes with code blocks, trying to figure it out now. 

Both of the above are closely related.

**Comment by janogarcia** [comment]
> There's the expand code block button, but it doesn't work anymore due to jump-to-quote behavior. I think **we should replace the expanding code block in quotes with the expand button for quotes** by default, what do you think?

Do you mean that when I expand a collapsed quote that contains, let's say **some text and a code block** (not just a code block), we should expand both the text for that message and the code block itself?

If that's the case I wouldn't recommend doing it. Let's keep those actions separate. Otherwise the behavior could feel inconsistent, plus it creates some undesired coupling between those features (let's say we want to change the code block expand behavior to be shown on a modal instead of expanded inline).

My recommendation would be to hide the code block inline actions when shown in the context of a collapsed quote, and then reveal those actions when the quote is expanded.

**Comment by Palid** [comment]
I'm leaving hiding/showing actions for later in case of quoted codeblock, as doing it requires writing some hacky code that I'd prefer to not write; we should refactor the way we're rendering things to make it simpler instead so that we won't have weird state race conditions or invalid state between components.

**Comment by Palid** [comment]
@janogarcia I don't think you'll be entirely happy of this implementation, as I skipped some things due to time constraints, list of changes below:
1: The tooltip for hotkey isn't differently colored, as the buttons don't handle that yet and adding it would be out of scope for now. It's a very good addition though and we should definitely figure this out in some spare time.
2: I haven't added the floating menu bar, time constraints again.
3: I tried changing to suggested `Expand replies` but that sounds like expanding a thread of replies, and here we're expanding a quote. Changed the text to `Expand/Collapse quote` instead. Hoping it's fine, but please double check to make sure.
4: Added polish translation too.
5: There's a todo to show the button for expanding `code` block within a quote. Need to figure out with @dbkr how to make it well first, as I don't want to overcomplicate things to show one button.

TL;DR: it's nowhere as close as it should be in my opinion, but should be _good enough_ to try it.
Not sure how I feel about adding the hotkey tooltip without having it being properly supported too, though.

**Comment by novocaine** [comment]
@Palid can you post screenshots please

**Comment by janogarcia** [comment]
@Palid Thanks for the detailed update (and the hard work)!

1. Yeah, let's figure out how to extend the tooltip component properly in a separate issue.
2. That wasn't actually a requirement. 😊👍 
3. Makes sense. My only suggestion was updating it to `Expand/Collapse quotes` (plural), which you've already addressed in the latest commit.
4. Dziękuję!
5. Let's move all the improvement ideas and pending stuff over to a new issue.

I think we're already addressing all the core issues that were introduced with the non-expandable truncated quotes, so this should be good to go! Thanks, Palid!
